### PR TITLE
Add authentication service

### PR DESCRIPTION
**Pull request by SHtress**

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -0,0 +1,6 @@
+fastapi
+pydantic
+jwt
```
Comment (AlexandrZagorskiy) [inline requirements.txt]: change to PyJWT

File: requirements.txt [hunk 2]
```diff
@@ -0,0 +1,6 @@
...
+pyOpenSSL
+sqlalchemy
+
```
Comment (AlexandrZagorskiy) [inline requirements.txt]: Add uvicorn

Comment (AlexandrZagorskiy) [reply requirements.txt]: I think better to add poetry to project

File: README.md [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+GET request-body: {"token": "..."}
+Current credentials to test: login - alex, password - a665a45920422f9d417e4867efdc4fb8a04a1f3fff1fa07e998e86f7f7a27ae3 (SHA256 hash)
+All added endpoints respond with specific message codes.
+There are eight of them
+cnt.MSG_ALL_DONE: 0,
+cnt.MSG_INVALID_USERNAME: 1,
+cnt.MSG_INVALID_PASSWORD: 2,
+cnt.MSG_USER_NOT_FOUND: 3,
+cnt.MSG_USER_IS_IN_BASE: 4
+
```
Comment (AlexandrZagorskiy) [inline README.md:40]: Update readme

File: admin.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+        message_desc = cnt.MSG_USER_IS_IN_BASE
+    else:
+        message_desc = cnt.MSG_ALL_DONE
+    return message_desc
+
+
+@router.post("/user", response_model=Response)
+async def create_user(user: UserInCreate):
+    user_info = user.dict()
+    print(user_info)
```
Comment (AlexandrZagorskiy) [inline admin.py]: remove prints, or change to loggings

File: admin.py [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+    database = DataBase()
+    msg_desc = _verify_user_info_in_database(
+            database,
+            name=user_info[cnt.NAME],
+            password=user_info[cnt.PASSWORD]
+            )
+    response = get_response_message(msg_desc)
+    if msg_desc == cnt.MSG_ALL_DONE:
+        database.add_user(user_info[cnt.NAME], user_info[cnt.PASSWORD])
+    sc_response = requests.post(url="http://127.0.0.1:8090/admin/user", json=user_info).json()
```
Comment (AlexandrZagorskiy) [inline admin.py]: move main url with routes to config file. I will send you example of config file from another project

File: auth.py [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+        cnt.ISS: params[cnt.ISSUER],
+        cnt.EXP: time.time() + life_span,
+        cnt.USERNAME: username
+    }
+    token = jwt.encode(payload, key=private_key, algorithm='RS256')
+    return token
+
+def _generate_keys() -> None:
+    pkey= crypto.PKey()
+    pkey.generate_key(type=crypto.TYPE_RSA, bits=2048)
```
Comment (AlexandrZagorskiy) [inline auth.py]: move "2048" to constants

**Review by AlexandrZagorskiy — CHANGES_REQUESTED** [review]

File: run_server.sh [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+  pp=${platform_port:-8090}
+  python3 -m uvicorn main:app --reload --host "$fi" --port "$fp" 
+}
+
+help()
+{
+   echo "i     ip address of a flask server (default value is 127.0.0.1)"
+   echo "p     port of a flask server (default value is 5000)"
+   echo "s     ip address of a platform sc-server (default value is 127.0.0.1)"
+   echo "c     port of a platform sc-server(default value is 8090)"
```
Comment (AlexandrZagorskiy) [inline run_server.sh]: remove unused args from script, update --help info

**Review by AlexandrZagorskiy — CHANGES_REQUESTED** [review]

File: constants.py [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+REFRESH_TOKEN = 'refresh_token'
+REFRESH_TOKEN_LIFE_SPAN = 'refresh_token_life_span'
+ISSUER = 'issuer'
+USERNAME = 'username'
+# paths
+SQLITE_DB_PATH = 'sqlite_db_path'
+PRIVATE_KEY_PATH = 'private_key'
+PUBLIC_KEY_PATH = 'public_key'
+# for new user generation
+WS_JSON_URL = 'ws://localhost:8090/ws_json' #TODO Add dynamic generation for url
```
Comment (AlexandrZagorskiy) [inline constants.py]: move to config, and use existed constants

File: admin.py [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+
+@router.get("/users")
+async def get_users(token: Token):
+    database = DataBase()
+    users = database.get_users()
+    log.info(users)
+    return users
+
+
+async def generate_user_args(user_info):
```
Comment (AlexandrZagorskiy) [inline admin.py]: why async?

File: admin.py [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+    const.create_link(sc_types.LINK, role_content, f'{cnt.ROLE}_{role}')
+
+    const.create_node(sc_types.NODE_CONST, f'{cnt.NODE}_{role}')
+
+    const.create_edge(sc_types.EDGE_ACCESS_CONST_POS_PERM, f'{cnt.NODE}_{role}', f'{cnt.EDGE}_{role}')
+
+    const.create_edge(sc_types.EDGE_D_COMMON_CONST, f'{cnt.ROLE}_{role}', f'{cnt.NODE}_{role}', f'{cnt.ROLE}_{cnt.EDGE}_{role}')
+
+    const.create_edge(sc_types.EDGE_ACCESS_CONST_POS_PERM, f'{cnt.ARGS}_{cnt.NODE}', f'{cnt.ROLE}_{cnt.EDGE}_{role}', f'{cnt.ARGS}_{role}_{cnt.EDGE}')
+    return const
```
Comment (AlexandrZagorskiy) [inline admin.py]: You can skip return statement here

File: admin.py (hunk 9 above)
Comment (AlexandrZagorskiy) [reply admin.py]: Also review typing in new functions

**Review by AlexandrZagorskiy — CHANGES_REQUESTED** [review]

File: sc-server-auth/server/server_tests.py [hunk 11]
```diff
@@ -0,0 +1,45 @@
+import unittest
+import requests
+from config import params, BASE_AUTH_SERVER_URL
+from server import constants as cnt
+
+user_creds = {"name": "admin", "password": "a665a45920422f9d417e4867efdc4fb8a04a1f3fff1fa07e998e86f7f7a27ae3"}
+wrong_user_creds = {"name": "stas", "password": "a665a45923422f9d417e5867efdc4fb8a06a1f3fff1fa07e998e86f7f7a27ae3"}
+wrong_pass_creds = {"name": "admin", "password": "a665a45923422f9d417e5867efdc4fb8a06a1f3fff1fa07e998e86f7f7a27ae3"}
```
Comment (FallenChromium) [inline sc-server-auth/server/server_tests.py:8]: I'm not quite sure why `user_creds` password and `wrong_pass_creds` password are same - is it intended?

File: sc-server-auth/server/admin.py [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+router = APIRouter(
+    prefix="/admin",
+    tags=["admin"],
+)
+
+
+def _verify_user_info_in_database(database: DataBase, name: str, password: str) -> str:
+    if not username_verifier.verify(name):
+        message_desc = cnt.MSG_INVALID_USERNAME
+    elif len(password.strip()) == 0:
```
Comment (FallenChromium) [inline sc-server-auth/server/admin.py:20]: That doesn't look right to me. Perhaps we could move password validation to a separate function and, while we're at it, create some additional requirements for the password (no non-latin characters, some minimal length, etc.)

File: sc-server-auth/server/server_tests.py (hunk 11 above)
Comment (SHtress) [reply sc-server-auth/server/server_tests.py:8]: Test will be changed latter, not so important for now

File: sc-server-auth/server/admin.py (hunk 12 above)
Comment (SHtress) [reply sc-server-auth/server/admin.py:20]: Yes it's good idea, we will think about later, but now we need working version without creating and deleting users

**Review by AlexandrZagorskiy — APPROVED** [review]

**Pull request merged**